compiler/mir/astgen.nim

```diff
@@ -248,14 +248,52 @@ proc wrapArg(stmts: seq[PNode], info: TLineInfo, val: PNode): PNode =
     result = newTreeIT(nkStmtListExpr, info, val.typ, stmts)
     result.add val
 
-func unwrap(expr: PNode, stmts: var seq[PNode]): PNode =
-  if expr.kind == nkStmtListExpr:
-    for i in 0..<expr.len-1:
-      stmts.add expr[i]
+proc flatten(expr: PNode, stmts: var seq[PNode]): PNode =
+  ## Applies the following transformations to the expression `expr`:
+  ## 1.) transform an AST like: ``(A (StmtListExpr (x) (y)) (z))`` into
+  ##     ``(StmtListExpr (x) (A (y) (z)))``
+  ## 2.) add all statements from a statement list expression to `stmts` and
+  ##     pass on the last node
+  ##
+  ## The result of #2 is passed back to step #1 and the process is repeated
+  ## until none of the two transformations apply anymore. Together, these
+  ## transformations make processing easier for the following analysis steps,
+  ## and the generated AST a bit less nested.
+  proc forward(n: PNode, p: int): PNode =
+    ## Performs transformation #1
+    if n[p].kind == nkStmtListExpr:
+      result = n[p]
+      n[p] = result[^1]
+      result[^1] = n
+    else:
+      result = n
 
-    result = expr[^1]
-  else:
-    result = expr
+  var it = expr
+  while true:
+    # we're looking for expression nodes that represent side-effect free
+    # operations
+    case it.kind
+    of nkDotExpr, nkCheckedFieldExpr, nkBracketExpr, nkHiddenAddr, nkAddr,
+      nkDerefExpr, nkHiddenDeref, nkCStringToString, nkStringToCString,
+      nkObjDownConv, nkObjUpConv:
+      it = forward(it, 0)
+    of nkConv, nkHiddenStdConv, nkCast:
+      it = forward(it, 1)
+    else:
+      # no AST to which transform #1 applies
+      discard
+
+    if it.kind == nkStmtListExpr:
```

Comment:
(not sure if I'd make this change, you might have already thought of it)

technically this could join the `case` statement above? And assuming the consts existed, the while loop predicate could be `it.kind notin nkSideEffectFreeKinds + nkCastAndConvKinds + {nkStmtList}`?

Not sure if that's necessarily easier to reason about, though.

Reply:
Yep, I have, and ultimately decided against joining both statements. The extra `if` does stand a bit out, but I think the step-by-step aspect makes it a tiny bit more clearer.